### ISSUE TITLE
Embedding API: How to get Function arguments other then bool, double and string?

### ISSUE THREAD

**Issue by verpeteren** [munificent/wren#180]
# Introduction

I am still in the process of evaluating wren to be used in a project as the embedded scripting language.

# Research

So far I know how to use `wrenGetArgumentBool`,`wrenGetArgumentDouble`, `wrenGetArgumentString`, but other types might be usefull as well. I read in https://github.com/munificent/wren/issues/156 that the internal types (I guess that means: list, map, class, function) should not be exposed in the Wren's public API. That seems a sound decision. 

# Scripted example

The following is aperantly working correcty for a class instance, but I can image also variants with maps, lists and functions: 

```
class Person{
	new( name ){
		__lastName = "doe"
		_name = name
	}
	name { 
		return _name
	 }
	meetInstance( /*class Person*/ otherPerson ) {
		IO.print( "hi " + otherPerson.name + "! I am " + _name + " " +__lastName )
	}
	meetString( /*string*/ otherPersonName ) {
		IO.print( "hi " + otherPersonName + "! I am " + _name + " " +__lastName )
	}
}

var pete = new Person( "pete" )
var john = new Person( "john" )
john.meetInstance( pete )
john.meetString( pete.name )
```

# Question

I guess that if it is working in a script, it should also be possible to achieve the same with the C API.
Please advice what would be the preferred way?

# TLDR;

The C API documents how to get bool, double and string arguments in a function, but it is unclear how to read other types (list, map, class, function).

**Comment by munificent** [comment]
> I guess that if it is working in a script, it should also be possible to achieve the same with the C API.

Not necessarily. Wren has some limitations in what C functions can do. Part of it is because we want to minimize the number of implementation details that leak into the C API. Also, because of how fibers are implemented, A C function cannot invoke code written in C. C functions can only be "leaves" in the call tree.

For cases where you want to pass some aggregate object to C, my idea is that you'd do that disaggregation in Wren in some wrapper method which then calls the C method passing only primitive values.

I realize that may be too painful to do, though. But, before I have a better solution, what I need is concrete use cases. Since I haven't spent much time embedding Wren in an app yet, I don't have a good feel for where the C API is unpleasant to use, and I don't want to speculatively add features to it unless I know they'll be helpful.

I know you're still evaluating things, but do you have any more details on what you're trying to accomplish with the API yet?

**Comment by verpeteren** [comment]
Hmm that troubles me. I have 2 use case in mind where I think that I would like to use wren:

### Case 1: webserver, scriptable with wren instead of php/python/javascript

One of my projects is a async-non-blocking server a-la node/vert.x. Sofar, a basic webserver, webclient, and database connections to mysql and postgresql are working. I want that these events can be scripted. Currently, I am using mozilla's spidermonkey javascript interpreter.
Basically I want to do the same with wren as I did in https://github.com/verpeteren/urbin/blob/master/src/glot/javascript.cpp but it the gc-rooting in spidermonkey is quite cumbersome.

### Case 2: Dynamic evaluating properties of data records

The other project is a Game-Theory calculation for the Deferred Acceptance Algorithm. Say I have: 

* a wren file with:

 ** several "student" classes, with one method to rate each school
 ** several "school" classes, with one method to rate each student
 **  a special 'main' function or class where 2 csv files are 'named' and one variabable is set.

* a csv file with schools  (approx: 7000 records): id, name, a 'json'string with properties, the name of the school class for this record 

* a csv with students  (approx: 34000 records): id, name, a 'json'string with properties, the name of the student class for this record.

The c program load the csv files:
* For every record in the csv file, id, the name and the 'json' string with properties will be fed in the constructor of the appropiate class. 
* every school rates every student 
* every student rates every school
* the deferred acceptance algorithm kicks in and spits out the best stable mariage proposals
* ..
*  profit

In the example it is schools&student, but the c-program does not care what parties are involved ( medical residents & hospitals, factory orders & production equipment, ..) the trick is that the enduser can mock up his rating method for each class, and tell each record, what class it belongs to. 

### TLDR;
This all probably sounds very abstract; I can give more details in a private email if you want.

Wren looks very promising, as it is small, fast, gotcha-less and understandble for non-programmers. 
I am looking forward to your reaction.

**Comment by munificent** [comment]
> One of my projects is a async-non-blocking server a-la node/vert.x. Sofar, a basic webserver, webclient, and database connections to mysql and postgresql are working.

This sounds like a perfect use case for Wren, and I'd be more than happy to hack on the embedding API to help make this go.

If it seems like Wren isn't making this easy, it's not because I don't *want* to, it's just that you're the first person to show up and actually kick the tires. There's a lot of work to be done. :smile:

> Basically I want to do the same with wren as I did in https://github.com/verpeteren/urbin/blob/master/src/glot/javascript.cpp

Cool, having a concrete use case like this is very helpful.

> the gc-rooting in spidermonkey is quite cumbersome.

Yup. This is one of the biggest challenges in an embedding API between a managed language like Wren or JS and C. A lot of Wren's current limitations come from this: by only exposing simple values, it ensures the GC can trace through every object on the heap without having references to objects hiding over in the C API.

Looking at javascript.cpp, it looks like you want it to be async (which is great). In Wren, I think the ideal way to do that is using fibers. That way you get a sync *API*, but other fibers can run while we're waiting for the networking stack to respond.

Of course, this means we'd need a little fiber scheduler under the hood. You'd write this mostly in Wren. Any time you do an operation that blocks on IO, you'd tell the scheduler what IO operation you're blocking on and it would store something like a map of IO operations to fibers.

Most of the bare pieces exist in Wren to do this, but it needs some more work. The rough idea is that when you want to have an "async" method in Wren, it becomes a pair of interop calls:

1. First, the Wren code calls a foreign function to begin the operation. That synchronously returns some identifier that's used to associate the current fiber with the operation its waiting on. Something kind of like:

    ```dart
    class OS {
      static getHostByName(name) {
        var id = foreignGetHostByName(name)
        __fibers[id] = Fiber.current
        // Suspend current fiber...
        // Get result value when fiber is resumed...
        return address
      }
    }
    ```

    Where the foreign function is just:

    ```c
    static void foreignGetHostByName(WrenVM* vm) {
      const char* address = wrenGetArgumentString(vm, 1);

      int id = MakeUniqueId();
      Core_GetHostByName(address, DnsResultHandler, (void *)id);

      wrenReturnDouble(vm, id);
    }
    ```

    After you call this from Wren, the current fiber gets suspended (*somehow*&mdash;this is a missing piece in Wren) and control returns back to the host application. At this point the Wren VM is idle, or possibly running other fibers if other ones are not also blocked waiting on IO.

2. Eventually, the asynchronous operation completes and that C callback is invoked. That gets the ID and the resulting data and calls a method in Wren to pass that in. Wren uses the ID to find the fiber that was blocked, and resumes it with the resulting data. Something kind of like:

    ```dart
    class OS {
      static completeGetHostByName_(id, ip) {
        var fiber = __parkedFibers[id]
        fiber.run(ip)
      }
    }
    ```

What do you think? Does this sound like a promising direction?

**Comment by verpeteren** [comment]
Thank you for your swift reply.
I am glad that you agree that my use-case is a correct scenario for wren.

Your answer is quite extensive, with a lot of stuff that I do not understand yet. I need some time tho  wrap my head around it, but sofar:

1. The 'onComplete' callback for the Fiber makes sense
2. To handle the suspension of fibers, could a eventloop be used? 
3. Nevertheless, I also would need to store some private data ( see https://github.com/munificent/wren/issues/156 ), I do not know what should come first.
4. To handle function arguments, other then bool, double and string, (my original question). I would prefer 

    ```
    var postgresql = new Postgresql( {host : 'dbServer', db : 'theDb', user : 'dbUser, password : 'dbPassword', port : dbPort }, timeout );
    ```
    
    But I also could do:
    
    ```
    var postgresql = new Postgresql( 'dbServer', 'theDb', 'dbUser, 'dbPassword', dbPort, timeout );
    ```

5. And I still would need to create Lists/Maps in the C code because I would like to return a list of rows ( and a row is basically a map)

    ```
    postgresql.query( "SELECT name, sales FROM sales WHERE customer = $1 OR sales > $2 ;" , ['foobar', 1000 ], new Fn{|resultList|
         if ( resultList == null ) {
             IO.print( "error" )
        } else {
            for ( row in resultList ) {
                    IO.print( "name: " + row.name + ' sales: ' + row.sales );
            }
        }
     } );
    ```
    
    I guess that an alternative with bool, double and string would less convienient ( probably  something like):
    
    ```
    postgresql.query( "SELECT name, sales FROM sales WHERE customer = '" + postgreqsl.escapeString( 'foobar' ) + "'  OR sales > '" + 1000 + "', new Fn{|resultList|
         if ( resultList == null ) {
             IO.print( "error" )
        } else {
            var columnNames = []
            var colId = 0
            while ( ( colName = postgresql.getColumnName(resultList, colId)) ) {
                colId = colId + 1
                columnNames.add( colName )
            }
            var colId = 0
            var rowId = 0
            while ( ( rowId < postgresql.getRowCount( resultList ) ) ) {
                var value
                IO.print( "\n" )
                for ( colunmName in columnNames ) {
                    value = postgresql.getValue( resultList, rowId, colId )
                    IO.print( "\t" + columnName + ":" + value )
                    colId = colId + 1
                }   
                rowId = rowId + 1
            }
        }
    } );
    ```
    
    The first disadvantage is that we do not have the benefit of parameterised querys. But the code above is allot like a wrapper to the to postgresql's libpq functions. The mysql's mysac functions are different; Direct access to these function might be faster but I think that the back and forth jumping between wren and the c code in the loops hurts. This direct access feels php-ish (and I mean that in the negative way)


The more I think about wren, the more I like it. But I can understand that my vision does not align with wren's goals. I do not want to make things more complicated: I just want it to make it easier :-).
I am happy to mock something up if you tell me what direction to go.

**Comment by munificent** [comment]
> To handle the suspension of fibers, could a eventloop be used?

Yes, there will be some kind of event loop or scheduler going on. I've implemented something [very similar to that](https://github.com/munificent/magpie/blob/master/src/VM/Scheduler.cpp) in a previous language that also had fibers—though Magpie's fibers are a bit different. I just haven't had the time to try to prototype something like that for Wren.

My current thought is that you could do this at the application level. The core Wren VM shouldn't need an event loop.

> Nevertheless, I also would need to store some private data ( see #156 ), I do not know what should come first.

Yeah, that's definitely necessary. You could work around it by storing something like numeric IDs in the Wren object that refer to some sort of look-up table you manage in C code. But that still leaves you with the problem of not knowing when that data can be cleaned up.

> To handle function arguments, other then bool, double and string, (my original question). I would prefer
>
>     var postgresql = new Postgresql( {host : 'dbServer', db : 'theDb', user : 'dbUser, password : 'dbPassword', port : dbPort }, timeout );
>
> But I also could do:
>
>     var postgresql = new Postgresql( 'dbServer', 'theDb', 'dbUser, 'dbPassword', dbPort, timeout );

You could actually do this now, by adding a layer of wrapping:

```dart
class Postgresql {
  new (config, timeout) {
    _inner = new NativePostgresql(config["host"], config["db"], config["user"], config["password"], config["port"], timeout)
  }
}
```

That is a bit tedious, though. It would be nice if you could define constructors that didn't implicitly create an instance of the surrounding class, but could create and return whatever they wanted.

> I guess that an alternative with bool, double and string would less convienient ( probably something like): ...

Yeah, that's what I had in mind. It is a lot of back and forth, but it might be something to at least start with.

>  I think that the back and forth jumping between wren and the c code in the loops hurts.

The question is how much? The performance may not be bad to do this, in part *because* Wren's C API is so simple. Right now, it doesn't have to do very much work to prepare to call native code or after a foreign call is finished.

> But I can understand that my vision does not align with wren's goals. 

It does, I think. It will just take a while to get there. The C API is the least mature part of Wren right now, and your needs are fairly sophisticated. That's a good thing, it just means it might require a lot of patience on your part. :)

**Comment by MarcoLizza** [comment]
> And I still would need to create Lists/Maps in the C code because I would like to return a list of rows ( and a row is basically a map)

Pardon my intrusion, but I feel that the best way to face this problem would be to implement a host-side UDT.

Currently Wren does not permit that (see #156), but implementing a recordset-like abstraction type would be quite and easy and versatile solution (this do sound a bit like the old-fashioned ADO way to do it, but that not necessarily a bad thing).

**Comment by munificent** [comment]
Yeah, that would also work. If you want the query results to remain mostly on the C side of things, that's a good solution.

**Comment by munificent** [comment]
Howdy. I've done a lot of work on Wren and the embedding API since this was filed. Since then:

1. I added IO libraries to the CLI. This gave me a chance to figure out how to build an asynchronous fiber scheduler on top of fibers. That required adding a couple of methods to Fiber: [suspend()](http://munificent.github.io/wren/modules/core/fiber.html#fiber.suspend()) and [transfer()](http://munificent.github.io/wren/modules/core/fiber.html#transfer()). So now I know it is possible to write a scheduler on top of the VM.

2. A while back, I added support for "foreign classes". These give you a way to have a Wren object containing raw bytes that are managed from C. They also support finalizers.

3. I did a bunch of work revamping the C API, which just landed in master. I still need to write docs for it, but if you look at wren.h, there's a lot there.

Given those, how do you think we're looking?

**Comment by verpeteren** [comment]
Great!

I took another look at the embedding api, and there were indeed several things change since Imade my last test. So I took my [previous notes](https://groups.google.com/forum/#!topic/wren-lang/ES2If5zQYGg) and started a real-[toy](http://corewar.co.uk/standards/icws94.txt)-project that I still have on my bucket list.
I will use this as my reference, and let it evolve step by step, together with wren. :-)

My remarks on the current state of the api:

* I like the current 'slots' manner; I can see the benefits of it instead of the  --Arguments-- or --Return-- manner.
* It took my a while to find out why I had no output. RTFM on me; the 'writeFn' was not (yet) documented on the api-page of website.
* I even used the 'config.reallocateFn' with the [nspr memory functions] (https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSPR/Reference/Memory_Management_Operations#Memory_Allocation_Functions). That is realy cool! Nice!

The next thing that that I need to test/play with is the 'list' and the 'wrenvalue'.
It looks very practical, I hope that I can find time to look at this in more detail.

**Comment by munificent** [comment]
> RTFM on me; the 'writeFn' was not (yet) documented on the api-page of website.

Yeah, I have a *lot* of documenting to do now. :)

> I even used the 'config.reallocateFn' with the nspr memory functions. That is realy cool! Nice!

Very good to hear!

**Comment by munificent** [comment]
I'm closing this out for now since we've made a lot of progress on the API (though there's still lots of documenting and other work left to do). I'm sure there are still more deficiencies in the C API, but it will probably be easier to track them by opening new more targeted bugs for each thing.
